frontend/userpatch.lua

```diff
@@ -0,0 +1,121 @@
+local logger = require("logger")
+local DataStorage = require("datastorage")
+
+-- the directory KOReader is installed in (and runs from)
+local package_dir = lfs.currentdir()
+-- the directory where KOReader stores user data
+local data_dir = DataStorage:getDataDir()
+
+--- Run lua patches
+-- Execution order order is alphanum-sort for humans version 4: `1-patch.lua` is executed before `10-patch.lua`
+-- (see http://notebook.kulchenko.com/algorithms/alphanumeric-natural-sorting-for-humans-in-lua)
+-- string directory ... to scan through (flat no recursion)
+-- string priority ... only files starting with `priority` followed by digits and a '-' will be processed.
+-- return true if a patch was executed
+local function runUserPatchTasks(dir, priority)
+    if lfs.attributes(dir, "mode") ~= "directory" then
+        return
+    end
+
+    local patches = {}
+    for entry in lfs.dir(dir) do
+        local mode = lfs.attributes(dir .. "/" .. entry, "mode")
+        if entry and mode == "file" and entry:match("^" .. priority .. "%d*%-") then
+            table.insert(patches, entry)
+        end
+    end
+
+    if #patches == 0 then
+        return -- nothing to do
+    end
+
+    local function addLeadingZeroes(d)
+        local dec, n = string.match(d, "(%.?)0*(.+)")
+        return #dec > 0 and ("%.12f"):format(d) or ("%s%03d%s"):format(dec, #n, n)
+    end
+    local function sorting(a, b)
+        return tostring(a):gsub("%.?%d+", addLeadingZeroes)..("%3d"):format(#b)
+            < tostring(b):gsub("%.?%d+", addLeadingZeroes)..("%3d"):format(#a)
+    end
+
+    table.sort(patches, sorting)
+
+    for i, entry in ipairs(patches) do
+        local fullpath = dir .. "/" .. entry
+        if lfs.attributes(fullpath, "mode") == "file" then
+            if fullpath:match("%.lua$") then -- execute patch-files first
+                logger.info("User patch apply:", fullpath)
+                local ok, err = pcall(dofile, fullpath)
+                if not ok then
+                    logger.warn("User patch error applying patch:", fullpath, err)
+                    -- Only show InfoMessage, when UIManager is working
+                    if priority >= userpatch.late and priority < userpatch.before_exit then
+                        -- Only developers (advanced users) will use this mechanism.
+                        -- A warning on a patch failure after an OTA update will simplify troubleshooting.
+                        local UIManager = require("ui/uimanager")
+                        local InfoMessage = require("ui/widget/infomessage")
+                        UIManager:show(InfoMessage:new{text = "Error applying patch:\n" .. fullpath}) -- no translate
+                    end
+                end
+            end
+        end
+    end
+    return true
+end
+
+--- This function applies lua patches from `/koreader/userpatches`
+---- @string priority ... one of "early\_once", "early", "late", "before\_exit", "on\_exit"
```

Comment:
The escaping was weird, but, more importantly, this is outdated, AFAIK.

What needs to be passed is the numerical prefix, as dictated by the `userpatch` hashtable ;).

Reply:
The escae was due to markup, but anyway a hint that allowed values are the attributes of userpatch is easier to maintain +1